### ISSUE TITLE
Add support for (custom) external Markdown renderer

### ISSUE THREAD

**Issue by nikcorg** [revolunet/sublimetext-markdown-preview#284]
TL;DR Any chance of adding configurable external renderer support to the plugin?

I'd like to use [CommonMark](http://commonmark.org/) for rendering Markdown previews. Sadly, its [Python implementation](https://github.com/jgm/CommonMark/wiki/List-of-CommonMark-Implementations#python) is apparently out-of-date, so it probably doesn't make sense to rely on it. However there are several up-to-date implementations that could be used as external renderers, if that was an option.

I had a look at the plugin code, and I was able to monkey-patch in CommonMark support, but my Python-fu is simply not good enough to accomplish a Pull Request for a generic configurable renderer.

**Comment by facelessuser** [comment]
What kind of configuration would you expect from CommonMark?  As I understand, CommonMark has been extremely specific about their rules as they have spent a long time splitting hairs over every minute detail, so I am not sure what there would be to configure as configuring would alter the behavior.  Maybe I don't know enough about CommonMark.  Though I appreciate what they are trying to accomplish,  I don't use CommonMark  as I enjoy using Python Markdown as I can alter and bend it to what I want it to do.  I don't know what would be desired, which is why a CommonMark user would need to outline what they would want beyond what is already available.

At this point I am just curious what you would want beyond sending the markdown through CommonMark.  If **I** were to add it, I would literally drop it in as MultiMarkdown support as been dropped in; someone else may be more ambitious with their support or even me if I understood more what else could be configured.

**Comment by nikcorg** [comment]
Oh, I would be happy just to be able to set the command to send markdown
through. Although, a syntax such as

`"dialect":"cmd --flag"`

would probably cover all bases, or at least 99%.

.n

**Comment by nikcorg** [comment]
Now that I thought about it for a few more seconds, the above should
absolutely cover all bases, as long as the command name is configurable.

The reason I'm eager to use CommonMark is its pluggability. This is (at
least in my case) done via a custom script, which hooks into CommonMark for
the AST-parser, and finally the HTML-renderer, applying modifiers in
between.

So making just the command configurable would probably suffice for my use
case, including flags should cover anything else.

.n

Ps. Sorry for noise

**Comment by facelessuser** [comment]
So you just want to control the command call and all the flags and such?  Can you point me at the documentation for the command and its arguments?

I am assuming this is all for the C parser of CommonMark?

**Comment by nikcorg** [comment]
In my case I'm in fact talking about the JavaScript parser. Afaik it has no flags, and the command is simply "[commonmark](https://www.npmjs.com/package/commonmark)", no flags at all. For my own variants -which really just apply transformations such as header offset- there are currently no flags either, but can't say whether it will remain like that or not.

**Comment by facelessuser** [comment]
As there seems to be multi-variants of implementations that may all return things slightly different and take slightly different inputs, I am inclined to just create a generic Markdown option that you can configure with CommonMark or anything else.  I think it would be bad to specifically favor the JavaScript implementation as there would inevitably be requests for Ruby, Perl, Go, etc.  This way we can just drop in support for pointing at a binary and configure the flags.  It would be up to the user to figure out what parameters worked well for them and for the Markdown Preview  environment.  It is possible that MultiMarkdown support could be moved into this generic support as well.

**Comment by nikcorg** [comment]
I would absolutely advise against maid anything special. Pipes or nothing, I say! :-)

**Comment by facelessuser** [comment]
I may took a look at this in the next couple of weeks.

**Comment by facelessuser** [comment]
Assuming it gets merged, #304 should address this.

**Comment by nikcorg** [comment]
Fantastic!

**Comment by facelessuser** [comment]
Released in [`1.3.0`](https://github.com/revolunet/sublimetext-markdown-preview/releases/tag/1.3.0).  Read the readme for more instructions.  If you find issues, just open a new issue.
